File: Tests/SignalR.Extras.Autofac.Test/TestHelper.fs
namespace SignalR.Extras.Autofac.Test

open Autofac
open Autofac.Integration.SignalR
open Microsoft.AspNet.SignalR.Hubs
open SignalR.Extras.Autofac
open SignalR.Extras.Autofac.Test.Stubs
open SignalR.Extras.Autofac.Test.Stubs.Dependencies

[<AbstractClass; Sealed>]
type internal TestHelper private() =
    static member SetupAutofacContainer() =
        let builder = ContainerBuilder()
        builder.RegisterLifetimeHubManager()
        //Register stub dependencies for the hubs
        builder.RegisterType<ScopedObjectStub>().InstancePerLifetimeScope() |> ignore
        builder.RegisterType<SingletonObjectStub>().SingleInstance() |> ignore
        //Register our LifetimeHubs
        builder.RegisterType<DynamicLifetimeHubStub>().ExternallyOwned() |> ignore
        builder.RegisterType<GenericLifetimeHubStub>().ExternallyOwned() |> ignore
        //Register ordinary, untracked SignalR hubs
        builder.RegisterType<DynamicOrdinaryHubStub>().ExternallyOwned() |> ignore
        builder.RegisterType<GenericOrdinaryHubStub>().ExternallyOwned() |> ignore
        builder.Build()
    static member SetupSignalRHubManager(container: IContainer) =
        //Configure SignalR to use Autofac for dependency resolution
        let resolver = new AutofacDependencyResolver(container)
        //Note: SignalR's DefaultHubManager below is unrelated to our LifetimeHubManager
        DefaultHubManager(resolver)

File: Tests/SignalR.Extras.Autofac.Tests/TestHelper.cs
using Autofac;
using Autofac.Integration.SignalR;

using Microsoft.AspNet.SignalR.Hubs;

using SignalR.Extras.Autofac.Tests.Stubs;
using SignalR.Extras.Autofac.Tests.Stubs.Dependencies;


namespace SignalR.Extras.Autofac.Tests;

internal static class TestHelper
{

    internal static IContainer SetupAutofacContainer()
    {
        var builder = new ContainerBuilder();
        builder.RegisterLifetimeHubManager();

        //Register stub dependencies for the hubs
        builder.RegisterType<ScopedObjectStub>().InstancePerLifetimeScope();
        builder.RegisterType<SingletonObjectStub>().SingleInstance();
        builder.RegisterType<RequestScopedObjectStub>().InstancePerRequest(ScopeLifetimeTag.RequestLifetimeScopeTag);

        //Register our LifetimeHubs
        builder.RegisterType<DynamicLifetimeHubStub>().ExternallyOwned();
        builder.RegisterType<GenericLifetimeHubStub>().ExternallyOwned();

        //Register ordinary, untracked SignalR hubs  
        builder.RegisterType<DynamicOrdinaryHubStub>().ExternallyOwned();
        builder.RegisterType<GenericOrdinaryHubStub>().ExternallyOwned();

        return builder.Build();
    }


    internal static IHubManager SetupSignalRHubManager(IContainer container)
    {
        //Configure SignalR to use Autofac for dependency resolution
        var resolver = new AutofacDependencyResolver(container);

        //Note: SignalR's DefaultHubManager below is unrelated to our LifetimeHubManager
        return new DefaultHubManager(resolver);
    }

}